skrub/_sentence_encoder.py

```diff
@@ -0,0 +1,300 @@
+        batch_size=32,
+        use_auth_token=None,
+        cache_folder=None,
+        random_state=None,
+        verbose=False,
+    ):
+        self.model_name = model_name
+        self.n_components = n_components
+        self.device = device
+        self.batch_size = batch_size
+        self.use_auth_token = use_auth_token
+        self.cache_folder = cache_folder
+        self.random_state = random_state
+        self.verbose = verbose
+
+    def fit_transform(self, column, y=None):
+        """Fit the SentenceEncoder from ``column``.
+
+        In practice, it loads the pre-trained model from disk and returns
+        the embeddings of the column.
+
+        Parameters
+        ----------
+        column : pandas or polars Series of shape (n_samples,)
+            The string column to compute embeddings from.
+
+        y : None
+            Unused. Here for compatibility with scikit-learn.
+
+        Returns
+        -------
+        X_out : pandas or polars DataFrame of shape (n_samples, n_components)
+            The embedding representation of the input.
+        """
+        st = import_optional_dependency("sentence_transformers")
+
+        if not sbd.is_string(column):
+            raise RejectColumn(f"Column {sbd.name(column)!r} does not contain strings.")
+
+        self._check_params()
+
+        self.cache_folder_ = get_data_dir(
+            name=self.model_name, data_home=self.cache_folder
+        )
+        try:
+            self.estimator_ = st.SentenceTransformer(
+                self.model_name,
+                device=self.device,
+                cache_folder=self.cache_folder_,
+                use_auth_token=self.use_auth_token,
+            )
+        except OSError as e:
+            raise ModelNotFound(
+                f"{self.model_name} is not a local folder and is not a valid "
+                "model identifier listed on 'https://huggingface.co/models'.\n "
+                "If this is a private repository, make sure to pass a token having "
+                "permission to this repo by passing `use_auth_token=<your_token>`"
+            ) from e
+
+        self.input_name_ = sbd.name(column)
+
+        X_out = self._vectorize(column)
+
+        if self.n_components is not None:
+            if (min_shape := min(X_out.shape)) >= self.n_components:
+                self.pca_ = PCA(
+                    n_components=self.n_components,
+                    copy=False,
+                    random_state=self.random_state,
+                )
+                X_out = self.pca_.fit_transform(X_out)
+            else:
+                warnings.warn(
+                    f"The matrix shape is {(X_out.shape)}, and its minimum is "
+                    f"{min_shape}, which is too small to fit a PCA with "
+                    f"n_components={self.n_components}. "
+                    "The embeddings will be truncated by keeping the first "
+                    f"{self.n_components} dimensions instead. "
```

Comment:
I wonder about the choice of truncating in this case, have you seen or used that approach before? I guess the default behavior of scikit-learn pca would be to set the number of components to n rows in this case?

Reply:
Yes, SentenceTransformers proposes [this option](https://github.com/UKPLab/sentence-transformers/blob/master/sentence_transformers/util.py#L314) directly in [the encode method](https://github.com/UKPLab/sentence-transformers/blob/master/sentence_transformers/SentenceTransformer.py#L605).

But your question is interesting; my POV is that several arguments can back truncation when the number of rows is insufficient to fit a PCA (< 30, by default, which is fairly low):
1. Embedding dimensions are different across models, setting a default truncation would allow fixing the number of columns of the output.
2. Embedding dimensions are often huge (> 128, up to > 1000) which is tedious to deal with in the context of a dataframe.
3. Users can keep all dimensions with `n_components=None`, but the opposite wouldn't be true; users couldn't limit the number of embedding dimensions on a small dataset if we skipped the dimension reduction.
4. Limiting the number of components rather than using a PCA can even be advantageous with special encoders like [Matryoshka Embeddings](https://sbert.net/examples/training/matryoshka/README.html).

WDYT?

Reply:
ok in this case let's keep truncation :) the alternative I had in mind was not to keep all dimensions, but to set the number of components to min(n rows, n columns) with a warning so that the PCA could be fitted

Reply:
Ah ok! This alternative also makes sense. @GaelVaroquaux WDYT?

Reply:
I think that this case will be very infrequent, no? It fires when the number of components requested is smaller than the number of samples, no? And we will typically use as a number of components 30.
In which case it doesn't matter much which way we go.
But maybe I got confused

Reply:
So, when the number of samples is smaller than the PCA dimension (`n_components`, which defaults to 30), we have two options for reducing dimensionality:

1. Truncate the embeddings by selecting the first `n_components` dimensions
2. Adjust the `n_components` parameter of the PCA to match `n_samples` and then run the PCA

This situation often arises with very small or demo datasets.